### ISSUE TITLE
Different behavior between Spring Boot 3.0.5 & 2.7.10, when using @EntityGraph with @ElementCollection/@CollectionTable

### ISSUE THREAD

**Issue by Eng-Fouad** [spring-projects/spring-data-jpa#2894]
I have this field inside `UserEntity`:

    @ElementCollection(fetch = FetchType.LAZY)
    @CollectionTable(name = "users_roles_mapping", joinColumns = @JoinColumn(name = "user_id", referencedColumnName = "id"))
    @Enumerated(EnumType.ORDINAL)
    @Column(name = "role_id")
    Set<Role> roles;

and I have this method inside `UsersRepository`:

    @EntityGraph(attributePaths = {"roles"})
    Optional<User> queryFirstById(int id);

On Spring Boot v3.0.5, `user.getRoles()` always contain single role despite the user has multiple roles. On Spring Boot v2.7.10, `user.getRoles()` works as expected and it contains all assigned roles.

---

Here is a reproducer: https://github.com/Eng-Fouad/spring-entity-graph-bug

**Comment by schauder** [comment]
This sounds very much like a Hibernate issue, since all the JPA mapping is done by it.
I recommend creating a reproducer based purely on Hibernate and submitting an issue with them.

If you think this is actually a Spring Data issue, please provide a reproducer for the issue including a test case that demonstrates the correct behaviour using the `EntityManager` and the Criteria API directly.

**Comment by Eng-Fouad** [comment]
@schauder What Hibernate versions that Spring Boot 3.0.5 & 2.7.10 use?

**Comment by Eng-Fouad** [comment]
I couldn't figure how to replicate the issue using JPA API only.

Using Spring Boot 3.0.5, and given:

```
@Repository
public interface UsersRepository extends JpaRepository<UserEntity, Integer> {
	@EntityGraph(attributePaths = {"roles"})
	UserEntity queryFirstById(int id);
}

@RequiredArgsConstructor
@Slf4j
@Service
public class UsersService {

	private final UsersRepository usersRepository;

	@PersistenceContext
	private EntityManager entityManager;

	@Transactional(isolation = Isolation.READ_COMMITTED, readOnly = true)
	public Set<Role> getUserRolesByIdV1(int id) {
        return usersRepository.queryFirstById(id).getRoles();
	}

	@Transactional(isolation = Isolation.READ_COMMITTED, readOnly = true)
	public Set<Role> getUserRolesByIdV2(int id) {
		var cb = entityManager.getCriteriaBuilder();
		var criteriaQuery = cb.createQuery(UserEntity.class);
		Root<UserEntity> root = criteriaQuery.from(UserEntity.class);
		criteriaQuery.where(cb.equal(root.<Integer>get("id"), id));
		TypedQuery<UserEntity> typedQuery = entityManager.createQuery(criteriaQuery);

		EntityGraph<UserEntity> entityGraph = entityManager.createEntityGraph(UserEntity.class);
		entityGraph.addAttributeNodes("roles");
		typedQuery.setHint(EntityGraphType.FETCH.getKey(), entityGraph);

		UserEntity userEntity = typedQuery.getSingleResult();
		return userEntity.getRoles();
	}
}
```

`getUserRolesByIdV2` returns 2 items while `getUserRolesByIdV1` returns 1 item.

How make `getUserRolesByIdV2` do the exact same thing as `getUserRolesByIdV1`? I need your help.

**Comment by Eng-Fouad** [comment]
I was able to replicate the bug using JPA API:

	EntityGraph<UserEntity> entityGraph = entityManager.createEntityGraph(UserEntity.class);
	entityGraph.addAttributeNodes("roles");
	
	Query query = entityManager.createQuery("SELECT u FROM UserEntity u WHERE u.id = 1");
	query.setMaxResults(1);
	query.setHint(EntityGraphType.FETCH.getKey(), entityGraph);
	
	UserEntity userEntity = (UserEntity) query.getSingleResult();
	return userEntity.getRoles();

This returns 2 roles in `2.7.10` and 1 role in `3.0.5`.

**Comment by Eng-Fouad** [comment]
Submitted a bug report to Hibernate team: https://hibernate.atlassian.net/browse/HHH-16408

**Comment by schauder** [comment]
Thanks for the confirmation.

**Comment by Eng-Fouad** [comment]
This bug should be fixed in Hibernate 6.2.0: https://hibernate.atlassian.net/jira/software/c/projects/HHH/issues/HHH-15964
